### ISSUE TITLE
"Export equals" syntax when used with Webpack and Babel

### ISSUE THREAD

**Issue by JakubJagoda** [DefinitelyTyped/DefinitelyTyped#8005]
Before we start, a quick word on "why would you use TS with Babel"
# Tl;dr for using Babel
TypeScript does not currently support full set of features when compiled to ES5. For..of loops [are compiled to regular for loops](https://github.com/Microsoft/TypeScript/pull/2207), async/await support for ES5/ES3 is [not yet finished](https://github.com/Microsoft/TypeScript/issues/1664) and planned to be delivered in version 2.0. So in order to have these working in all today's browsers and environments, one can just set the TS output to es6 and then pass it further to Babel, which will take care of the rest.

# Export equals
However, this is a problem with TypeScript's specific "export equals" syntax. If used, then import of this module has to be imported using `import * as something` way. Let's look at simple example:
```typescript
//a.d.ts
declare var A;
declare module 'A' {
   export = A;
}

//b.ts
import * as A from 'A';
```
this is weird for me, as the handbook says 
>The export = syntax specifies a single object that is exported from the module

 but if you try to use `import A from 'A';` syntax you'll get an error telling you that this module has no default export. But okay, we can just simply use the `import * as` syntax instead.

# Where's the problem
If we try to bundle the application using Webpack + TypeScript loader then everything is fine. If we also use Babel loader on the way (to cover the case I described in the first point), then it starts not be fine. Why? Because Babel handles `import * as` in a special way. It creates a plain empty object, and then fills it with all the exports as properties. **This causes different problems with various libraries**. So far I identified 3 of them (Bluebird, jQuery, d3), but I suppose there's more.

## Bluebird, jQuery and probably other callables
With Bluebird and jQuery the main problem is that what is exported are functions. When it gets transpiled by Babel, these functions get replaced by objects. The objects has all the properties the functions had, but are not callable anymore, which makes it impossible to actually use these two libraries (a hack exists, but it doesn't apply to all the libs. I'll describe it later).

## d3
With d3 the problem is a bit different. d3 does not export a callable, therefore we can still use it and everything seem to work fine. However, d3 uses a "global" place (`d3.event`) where it stores the object representing last triggered event. Let's not discuss if it's a bad idea and why it is, this is just the way this lib works. So, because of how Babel treats imports, `d3.event` that d3 uses internally and `d3.event` that we use in the module where we imported d3, do not refer to the same thing. Therefore, we are not able to access the information about triggered event.

## demo
Here is a repository that demonstrates the problem: https://github.com/JakubJagoda/dt-babel-problem
Just clone it and follow the README.

# How to solve it
All the problems I mentioned can be simply solved by changing the exports&import declaration from e.g.
```typescript
//export 
export = d3

//import 
import * as d3 from 'd3'
```
to:
```typescript
//export 
export default d3

//import 
import d3 from 'd3'
```
I think that conceptually it is reasonable to consider that these libraries export a single "thing" instead of a set of functions, as syntax `import * as x from 'x'` might suggest. Since it's an easy change I can open a PR which solves that at least for these 3 libraries (I expect there are more of them, but haven't spotted any yet) but first I want to know what the general opinion on that is. Maybe there is another way to solve this or any way to drop Babel but still using all TS features in ES5?

# Bonus: the hack
As I mentioned before, a simple hack exist (at least for callability problem), but it requires the library to have a function that returns the library object itself (for example jQuery has `noConflict()`). So we create a file, for example jquery-fix.ts and put this inside:
```typescript
import * as jQueryFactory from 'jquery';
const $ : JQueryStatic = <JQueryStatic>jQueryFactory.noConflict();
export default $;
```
and then, instead of importing `'jquery'` in our files, we import this module. Works for Bluebird too. However I'm not sure if a similar hack for d3-related problems exists, as well as for the libs that don't have functions returning the lib objects themselves.

**Comment by jesseschalken** [comment]
Does compiling with `"module": "commonjs"` and then sending it through Babel in CommonJS format work?

**Comment by JakubJagoda** [comment]
Yeah, indeed it seem to work. I needed to check it few times because I was sure I tried that before, but looks like changing `"module"` to `"commonjs"` makes it work correctly. Thanks, I hope this doesn't break anything else ;) I still consider `import jQuery from 'jquery'` a better-looking option than `import * as jQuery from 'jquery'`, but at least it works.

**Comment by vvakame** [comment]
Why do you post to this repo?
I think you should post to TypeScript repo.

**Comment by JakubJagoda** [comment]
@vvakame I proposed the solution in "How to solve it" section, did you read it?

**Comment by vvakame** [comment]
I hope to tsc become more closer to es6 too.
But I think `.d.ts` exists for tsc now. not with other transpiler.
I think you should talking about interop to other transpiler in TypeScript repo.

@DefinitelyTyped/contributors any comments?

**Comment by JakubJagoda** [comment]
Yeah, but I just thought it would be easier to change some typings definitions instead of proposing breakage of the whole `export = ` syntax behavior :) the decision that this would not compile down to a default export but rather to a collection of exported bindings (so like CommonJS module) was made almost a year ago, when ES6 modules were introduced to TypeScript. Sounds a bit too late for me to change that, therefore I decided to start raising this issue here.

**Comment by johnnyreilly** [comment]
Does this repo solve your problems: https://github.com/johnnyreilly/globalize-so-what-cha-want

I've only skim read but it sounds like you're trying to use TypeScript with Babel?  This repo does just that using the React / Flux definitions.

**Comment by JakubJagoda** [comment]
@johnnyreilly yes, I'm trying to use TS with Babel (for anyone who is curious why, I explained that at the very beginning, just in case). And actually the solution was super simple and it was enough just to change the module format to "commonjs" in TS. That's all and it works, so I'd say it's fair enough.

**Comment by johnnyreilly** [comment]
Cool - I'll close the issue.  I came upon this myself when trying to migrate a Babel project across to TypeScript.  Fortunately @jbrantly had my back with TS-Loader - but you know that now!

**Comment by nycdotnet** [comment]
FYI: In TypeScript 1.8 there is a new switch for tsc `--allowSyntheticDefaultImports` that works around this issue.

**What's new:**
https://github.com/Microsoft/TypeScript/wiki/What's-new-in-TypeScript#support-for-default-import-interop-with-systemjs
**Original Issue:**
https://github.com/Microsoft/TypeScript/issues/5285


If using TypeScript 1.8 with target `system` or if passing `--allowSyntheticDefaultImports`, an `export = foo` type definition file can be imported using `import foo from 'foo';`

**Comment by JakubJagoda** [comment]
That's great thing to know, @nycdotnet, thanks for sharing this :) I wasn't able to find that by myself.

**Comment by RReverser** [comment]
@JakubJagoda Also, just to clarify on:

> I still consider import jQuery from 'jquery' a better-looking option than import * as jQuery from 'jquery', but at least it works.

Yes, this might be nicer, but this is just hack that won't be compatible with any real ES6 module implementation, as spec says that only `default`-named export will be imported when using `import SomeName from` syntax. So it's anyway worth updating your code to be future-proof and compatible with upcoming implementations.

**Comment by jesseschalken** [comment]
@JakubJagoda `import jQuery = require('jquery')` works the same as `import * as jQuery from 'jquery'`, if you prefer that.

**Comment by RReverser** [comment]
@ jesseschalken Except that it's not ES6-compliant syntax either.

**Comment by jesseschalken** [comment]
So? Type annotations aren't ES6-complaint either.

The `import Foo from 'bar'` syntax is shared with ES6, and so it has to be ES6 compliant. (`--allowSyntheticDefaultImports` breaks this compliance, as you said.) The `import Foo = require('bar')` syntax isn't shared with ES6, so TypeScript can do what it wants with it.

**Comment by RReverser** [comment]
@jesseschalken IIUC, breaking ES6 compliancy with `export = ...` was what was the original issue about, as it couldn't be used with Babel, so using `import =` doesn't solve this. Even `allowSyntheticDefaultImports` is better in that sense because it at least can use the same syntax.

**Comment by JakubJagoda** [comment]
@RReverser 
>Yes, this might be nicer, but this is just hack that won't be compatible with any real ES6 module implementation, as spec says that only default-named export will be imported when using import SomeName from syntax

Of course, but importing it as `import * as jQuery from 'jquery'` is not spec-compatible either. This syntax will import all the exported bindings from 'jquery' module and attach them to **object** `jquery`. And as we all know, the core of jQuery is a **function**. So about that:
>So it's anyway worth updating your code to be future-proof and compatible with upcoming implementations.

if jQuery in its current state had to be rewritten to use ES6 modules, it could be done by either choosing a default module or a named binding (probably `jQuery`) pointing to the factory function. None of them plays nice with `import *` syntax (If latter was chosen then `import * as jquery from 'jquery'` would require us to use `jquery.jquery(selector)`. I'm not sure about the former, would it be available as `jquery.default`? Still bad anyway). So any syntax seems to be better and more future-proof than current, actually.
